.github/workflows/testing-big-endian.yml

```diff
@@ -7,32 +7,98 @@ on:
+
+      - if: ${{ steps.restore-cache.outputs.cache-hit != 'true' }}
+        run: |
+          set -x
+
+          mkdir -p /home/runner/linux-s390x-glibc
+
+          curl -L -o lastest.json https://api.github.com/repos/leleliu008/uppm-package-repository-linux-s390x-glibc/releases/latest
+
+          unset RELEASE_TAG_NAME_
+          unset RELEASE_FILENAMES
+
+          RELEASE_TAG_NAME_=$(jq -r '.tag_name'      lastest.json)
+          RELEASE_FILENAMES=$(jq -r '.assets[].name' lastest.json)
+
+          for filename in $RELEASE_FILENAMES
+          do
+              case $filename in
+                   jansson-*-linux-s390x-glibc.tar.xz) ;;
+                  libiconv-*-linux-s390x-glibc.tar.xz) ;;
+                   libxml2-*-linux-s390x-glibc.tar.xz) ;;
+                   libyaml-*-linux-s390x-glibc.tar.xz) ;;
+                     pcre2-*-linux-s390x-glibc.tar.xz) ;;
+                      zlib-*-linux-s390x-glibc.tar.xz) ;;
+                  *) continue ;;
+              esac
+              curl -LO https://github.com/leleliu008/uppm-package-repository-linux-s390x-glibc/releases/download/${RELEASE_TAG_NAME_}/${filename}
+              tar xf ${filename} -C /home/runner/linux-s390x-glibc --strip-components=1
+          done
+
+          rm -rf /home/runner/linux-s390x-glibc/share/log
+          rm -rf /home/runner/linux-s390x-glibc/share/man
+          rm -rf /home/runner/linux-s390x-glibc/share/doc
+          rm -rf /home/runner/linux-s390x-glibc/share/gtk-doc
+          rm -rf /home/runner/linux-s390x-glibc/lib/libz.so*
+          rm -rf /home/runner/linux-s390x-glibc/lib/lib*.la
+          rm -rf /home/runner/linux-s390x-glibc/installed-files
+          rm -rf /home/runner/linux-s390x-glibc/installed-metadata-ppkg
+
+      - run: sudo apt-get -y update
+      - run: sudo apt-get -y install automake autoconf pkg-config make gcc-s390x-linux-gnu
+
+      - run: ./autogen.sh
+      - run: |
+            export CC=/usr/bin/s390x-linux-gnu-gcc
+            export CPPFLAGS=-I/home/runner/linux-s390x-glibc/include
+            export LDFLAGS=-L/home/runner/linux-s390x-glibc/lib
+            export PKG_CONFIG_PATH=/home/runner/linux-s390x-glibc/lib/pkgconfig
+            ./configure --host=s390x-ibm-linux-gnu --enable-debugging || (cat config.log; exit 1)
+      - run: make -j$(nproc)
+
+      - run: /usr/bin/s390x-linux-gnu-readelf -h ctags
+      - run: /usr/bin/s390x-linux-gnu-readelf -d ctags
+
+      - run: |
+          cat > run.sh <<EOF
+          set -e
+
+          COLOR_GREEN='\033[0;32m'        # Green
+          COLOR_PURPLE='\033[0;35m'       # Purple
+          COLOR_OFF='\033[0m'             # Reset
+
+          echo() {
+              printf '%b\n' "\$*"
+          }
+
+          run() {
+              echo "\${COLOR_PURPLE}==>\${COLOR_OFF} \${COLOR_GREEN}\$@\${COLOR_OFF}"
+              eval "\$*"
+          }
+
+          run uname -a
+
+          run misc/units tmain --show-diff-output ./Tmain ./Tmain
```

Comment:
Why don't use "make units" and/or "make tmain" ?

Reply:
`make units` and `make tmain` would rebuild project. 

we cross-compile project on build-machine(ubuntu-x86_64) and do not want to rebuild it on QEMU(ubuntu-s390x).

Reply:
Oh, I see. How about extending Makefile handling such cases like:

```
diff --git a/makefiles/testing.mak b/makefiles/testing.mak
index c287c7160..f868b0627 100644
--- a/makefiles/testing.mak
+++ b/makefiles/testing.mak
@@ -11,7 +11,11 @@ check: tmain units tlib man-test check-genfile
 # clean-tlib and clean-gcov are not included
 clean-local: clean-units clean-tmain clean-man-test
 
+if TEST_WITHOUT_BUILDING
 CTAGS_TEST = ./ctags$(EXPECT)
+else
+CTAGS_TEST =
+fi
 READTAGS_TEST = ./readtags$(EXEEXT)
 MINI_GEANY_TEST = ./mini-geany$(EXPECT)
 OPTSCRIPT_TEST = ./optscript$(EXEEXT)
```

This is a pseudo code. I cannot remember the syntax representing the condition in Makefile quickly.

I want to hide the code behind "make units" and "make tmain" for people wanting to test ctags.
Your new code accesses the backyard. So I would like you to avoid it if possible.

Reply:
```
run apt-get -y update
run apt-get -y install make

run make tmain V=1 CI=1 SHELL=/usr/bin/bash PYTHON= CTAGS_TEST= READTAGS_TEST= OPTSCRIPT_TEST=
run make units V=1 CI=1 SHELL=/usr/bin/bash PYTHON= CTAGS_TEST= READTAGS_TEST= OPTSCRIPT_TEST=
```

this works, but need patch `misc/units`:

```diff
diff --git a/misc/units b/misc/units
index 0c952857..d062cd2a 100755
--- a/misc/units
+++ b/misc/units
@@ -1118,6 +1118,8 @@ action_run ()
 	ERROR 1 "No such directory(build_dir): ${build_dir}"
     fi
 
+    [ -z "$CTAGS" ] && CTAGS=./ctags
+
     if ! [ -f "${CTAGS}" ]; then
 	ERROR 1 "no such file: ${CTAGS}"
     elif ! [ -e "${CTAGS}" ]; then
@@ -1668,6 +1670,8 @@ action_fuzz_common ()
 	ERROR 0 "No such directory: ${units_dir}"
     fi
 
+    [ -z "$CTAGS" ] && CTAGS=./ctags
+
     if ! [ -f "${CTAGS}" ]; then
 	ERROR 1 "no such file: ${CTAGS}"
     elif ! [ -e "${CTAGS}" ]; then
@@ -2244,6 +2248,8 @@ action_tmain ()
 	ERROR 1 "No such directory(build_dir): ${build_dir}"
     fi
 
+    [ -z "$CTAGS" ] && CTAGS=./ctags
+
     if ! [ -f "${CTAGS}" ]; then
 	ERROR 1 "no such file: ${CTAGS}"
     elif ! [ -e "${CTAGS}" ]; then
```

Reply:
```
run make tmain V=1 CI=1 SHELL=/usr/bin/bash PYTHON= CTAGS_TEST= READTAGS_TEST= OPTSCRIPT_TEST=
```

This is not a good idea.

PYTHON= is really bad idea. Don't use the shell version anymore for running units and tmain. Let's use the script using python.

CTAGS_TEST= is hard to understand. "Passing value from an upper layer explicitly" is better than "getting value in lower layer" in this case.

```
make units TESTING_WITHOUT_BUILDING=1
```

is more understandable than

```
make CTAGS_TEST=
```
.

Reply:
With the following change, you can run "units" without (re)building ctags:

```
diff --git a/makefiles/testing.mak b/makefiles/testing.mak
index c287c7160..16ea00d9e 100644
--- a/makefiles/testing.mak
+++ b/makefiles/testing.mak
@@ -16,6 +16,11 @@ READTAGS_TEST = ./readtags$(EXEEXT)
 MINI_GEANY_TEST = ./mini-geany$(EXEEXT)
 OPTSCRIPT_TEST = ./optscript$(EXEEXT)
 
+CTAGS_TEST_DEP = $(CTAGS_TEST)
+READTAGS_TEST_DEP = $(READTAGS_TEST)
+MINI_GEANY_TEST_DEP = $(MINI_GEANY_TEST)
+OPTSCRIPT_TEST_DEP = $(OPTSCRIPT_TEST)
+
 if HAVE_TIMEOUT
 TIMEOUT = 1
 else
@@ -40,7 +45,7 @@ V_RUN_0 = @echo "  RUN      $@";
 #
 # SHELL must be dash or bash.
 #
-fuzz: $(CTAGS_TEST)
+fuzz: $(CTAGS_TEST_DEP)
 	$(V_RUN) \
 	if test -n "$${ZSH_VERSION+set}"; then set -o SH_WORD_SPLIT; fi; \
 	if test x$(VG) = x1; then		\
@@ -56,7 +61,7 @@ fuzz: $(CTAGS_TEST)
 #
 # NOISE Target
 #
-noise: $(CTAGS_TEST)
+noise: $(CTAGS_TEST_DEP)
 	$(V_RUN) \
 	if test -n "$${ZSH_VERSION+set}"; then set -o SH_WORD_SPLIT; fi; \
 	if test x$(VG) = x1; then		\
@@ -72,7 +77,7 @@ noise: $(CTAGS_TEST)
 #
 # CHOP Target
 #
-chop: $(CTAGS_TEST)
+chop: $(CTAGS_TEST_DEP)
 	$(V_RUN) \
 	if test -n "$${ZSH_VERSION+set}"; then set -o SH_WORD_SPLIT; fi; \
 	if test x$(VG) = x1; then		\
@@ -84,7 +89,7 @@ chop: $(CTAGS_TEST)
 		$${VALGRIND} --run-shrink \
 		--with-timeout=$(TIMEOUT)"; \
 	$(SHELL) $${c} $(srcdir)/Units
-slap: $(CTAGS_TEST)
+slap: $(CTAGS_TEST_DEP)
 	$(V_RUN) \
 	if test -n "$${ZSH_VERSION+set}"; then set -o SH_WORD_SPLIT; fi; \
 	if test x$(VG) = x1; then		\
@@ -100,7 +105,7 @@ slap: $(CTAGS_TEST)
 #
 # UNITS Target
 #
-units: $(CTAGS_TEST)
+units: $(CTAGS_TEST_DEP)
 	$(V_RUN) \
 	if test -n "$${ZSH_VERSION+set}"; then set -o SH_WORD_SPLIT; fi; \
 	if test x$(VG) = x1; then		\
@@ -161,7 +166,7 @@ validate-input:
 #
 # Test main part, not parsers
 #
-tmain: $(CTAGS_TEST) $(READTAGS_TEST) $(OPTSCRIPT_TEST)
+tmain: $(CTAGS_TEST_DEP) $(READTAGS_TEST_DEP) $(OPTSCRIPT_TEST_DEP)
 	$(V_RUN) \
 	if test -n "$${ZSH_VERSION+set}"; then set -o SH_WORD_SPLIT; fi; \
 	if test x$(VG) = x1; then		\
@@ -204,7 +209,7 @@ clean-tmain:
 		$(SHELL) $(srcdir)/misc/units clean-tmain $$(pwd)/Tmain; \
 	fi
 
-tlib: $(MINI_GEANY_TEST)
+tlib: $(MINI_GEANY_TEST_DEP)
 	$(V_RUN) \
 	builddir=$$(pwd); \
 	mkdir -p $${builddir}/misc; \
@@ -238,7 +243,7 @@ tinst:
 # Test readtags
 #
 if USE_READCMD
-roundtrip: $(READTAGS_TEST)
+roundtrip: $(READTAGS_TEST_DEP)
 	$(V_RUN) \
 	if ! test x$(CI) = x; then	\
 		ROUNDTRIP_FLAGS=--minitrip;			\
@@ -252,7 +257,7 @@ endif
 #
 # Checking code in ctags own rules
 #
-codecheck: $(CTAGS_TEST)
+codecheck: $(CTAGS_TEST_DEP)
 	$(V_RUN) $(SHELL) misc/src-check
 
 #
@@ -280,7 +285,7 @@ cppcheck:
 #
 # Testing examples in per-language man pages
 #
-man-test: $(CTAGS_TEST)
+man-test: $(CTAGS_TEST_DEP)
 	$(V_RUN) \
 	$(PYTHON) $(srcdir)/misc/man-test.py $(MAN_TEST_TMPDIR) $(CTAGS_TEST) $(srcdir)/man/ctags-lang-*.7.rst.in
 
```

```
% make units CTAGS_TEST_DEP=
  RUN      units
Traceback (most recent call last):
  File "/home/jet/var/ctags-github/./misc/units.py", line 1329, in <module>
    main()
  File "/home/jet/var/ctags-github/./misc/units.py", line 1326, in main
    sys.exit(func(subparser, *sys.argv[1:]))
...

I found we cannot use `if` in testing.mak freely.
Automake manages FOO of `if FOO`.

Reply:
I think both are understandable if epople do not read the source code. why do not use shell script to run tests?

Reply:
wrapped too much make makefike more complicated and run slower, also it is unfriendly to people who want to study or contribute to this project.

Reply:
The shell script (misc/units) was mostly obsoleted when @k-takata converted it to misc/units.py. We have maintained misc/units.py mainly since then. misc/units is in still our source tree because some minor parts of it are still used. 
Let's forget misc/units.


> wrapped too much make makefike more complicated and run slower, also it is unfriendly to people who want to study or contribute to this project. 

misc/units is MUCH slower and more complicated.
misc/units.py is faster but it is not simple enough because it was converted from misc/units.
misc/units.py is much larger and more complicated than testing.mak.

Reply:
I thought `misc/units.py` is for windows and `misc/units.py` is for `*nix` 😂

I think there need more comments in source code.

Reply:
`misc/units.py` is faster indeed.